### DISCUSSION TITLE
ControlNet v1.1.409 IP-Adapter error

### DISCUSSION THREAD

**Discussion by jaekwangkimho** [vladmandic/sdnext#2159] [Show and tell]
modules/modelloader.py in to 
def load_file_from_url(
    url: str,
    *,
    model_dir: str,
    progress: bool = True,
    file_name: str | None = None,
) -> str:
    """Download a file from `url` into `model_dir`, using the file present if possible.

    Returns the path to the downloaded file.
    """
    os.makedirs(model_dir, exist_ok=True)
    if not file_name:
        parts = urlparse(url)
        file_name = os.path.basename(parts.path)
    cached_file = os.path.abspath(os.path.join(model_dir, file_name))
    if not os.path.exists(cached_file):
        print(f'Downloading: "{url}" to {cached_file}\n')
        from torch.hub import download_url_to_file
        download_url_to_file(url, cached_file, progress=progress)
    return cached_file

**Comment by vladmandic** [comment]
what am i looking at hewe? code without any context?

**Reply by jaekwangkimho** [reply]
I couldn't write this long because my English was bad.  
ControlNet IP-Adapter function did not work, so I checked the error and found that it did not work because the corresponding module was missing.
So I left a post because I wanted the code to be added. I'm sorry I don't speak English.

**Reply by jaekwangkimho** [reply]
def load_file_from_url(
    url: str,
    *,
    model_dir: str,
    progress: bool = True,
    file_name: str | None = None,
) -> str:
    """Download a file from `url` into `model_dir`, using the file present if possible.Returns the path to the downloaded file."""
    os.makedirs(model_dir, exist_ok=True)
    if not file_name:
        parts = urlparse(url)
        file_name = os.path.basename(parts.path)
    cached_file = os.path.abspath(os.path.join(model_dir, file_name))
    if not os.path.exists(cached_file):
        print(f'Downloading: "{url}" to {cached_file}\n')
        from torch.hub import download_url_to_file
        download_url_to_file(url, cached_file, progress=progress)
    return cached_file
    
    I added this and it worked normally.

**Comment by vladmandic** [comment]
ok, i'll add it.
